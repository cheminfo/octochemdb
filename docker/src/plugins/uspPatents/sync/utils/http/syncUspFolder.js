import { join } from 'path';

import delay from 'delay';
import FSExtra from 'fs-extra';

import getFileIfNew from '../../../../../sync/http/utils/getFileIfNew.js';
import debugLibrary from '../../../../../utils/Debug.js';

import getFilesListUsp from './getFileListUsp.js';

const { mkdirpSync, existsSync } = FSExtra;

async function syncUspFolder(source, destinationFolder, year) {
  // create the destination folder if it doesn't exist
  const debug = debugLibrary('syncUspFolder');
  if (await !existsSync(destinationFolder)) {
    await mkdirpSync(destinationFolder);
  }
  // get the list of files to import
  let files = await getFilesListUsp(source, year, {
    fileFilter: (file) => file && file.name.endsWith('.zip'),
  });
  let filesDownloaded = [];
  if (process.env.NODE_ENV !== 'test') {
    for (let file of files) {
      let options = {
        filename: file.name.replace('.zip', ''),
        extension: 'zip',
      };
      let target = join(destinationFolder);
      let fileDownloaded = await getFileIfNew(file, target, options);
      if (fileDownloaded) {
        filesDownloaded.push(fileDownloaded);
      }
      debug(`Waiting 5 seconds before next download...`);
      await delay(5000);
    }
    return filesDownloaded;
  } else {
    // @ts-ignore
    filesDownloaded = Array.from(files);
    return filesDownloaded;
  }
}

export default syncUspFolder;
