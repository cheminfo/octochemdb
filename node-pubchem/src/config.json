{
  "mongodbUrl": "mongodb://localhost:27017",
  "databaseName": "pubchem",
  "dataFullDir": "./originalData/full",
  "dataWeeklyDir": "./originalData/weekly",
  "ftpServer": "ftp.ncbi.nlm.nih.gov"
}
