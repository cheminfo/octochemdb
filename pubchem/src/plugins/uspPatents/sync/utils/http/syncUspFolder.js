import { join } from 'path';

import { fileListFromPath } from 'filelist-from';
import FSExtra from 'fs-extra';

import getFile from '../../../../../sync/http/utils/getFile.js';

import getFilesListUsp from './getFileListUsp.js';

const { mkdirpSync, existsSync } = FSExtra;

async function syncUspFolder(source, destinationFolder, year) {
  // create the destination folder if it doesn't exist
  if (await !existsSync(destinationFolder)) {
    await mkdirpSync(destinationFolder);
  }
  // get the list of files to import
  const files = await getFilesListUsp(source, year);
  // get last file downloaded
  const lastFilesDownloaded = fileListFromPath(destinationFolder).sort(
    (a, b) => {
      a.path = a.webkitRelativePath;
      b.path = b.webkitRelativePath;
      if (a.path < b.path) return -1;
      if (a.path > b.path) return 1;
      return 0;
    },
  );

  // download the files
  const filesToDownload = files
    .filter((file) => {
      // skip until the last file downloaded
      //check for each file if it exists in the destination folder or if should be downloaded
      const fileName = file.name;
      const fileSize = file.size;
      file.path = join(destinationFolder, fileName);
      const fileExists = lastFilesDownloaded.find((lastFile) => {
        return lastFile.path === file.path && lastFile.size === fileSize;
      });
      if (fileExists) {
        return false;
      }
      return !fileExists;
    })
    .map((file) => {
      const fileName = file.name;
      const fileSize = file.size;
      file.path = join(destinationFolder, fileName);
      const fileExists = lastFilesDownloaded.find((lastFile) => {
        return lastFile.path === file.path && lastFile.size === fileSize;
      });
      if (fileExists) {
        return null;
      }
      return file;
    })
    .filter((file) => file);
  // download the files
  const filesDownloaded = await Promise.all(
    filesToDownload.map(async (file) => {
      const fileName = file.name;
      const fileSize = file.size;
      file.path = join(destinationFolder, fileName);
      const fileExists = lastFilesDownloaded.find((lastFile) => {
        return lastFile.path === file.path && lastFile.size === fileSize;
      });
      if (fileExists) {
        return null;
      }
      await getFile(file, file.path);
      return file;
    }),
  );
  // return the files downloaded
  return filesDownloaded;
}

export default syncUspFolder;
