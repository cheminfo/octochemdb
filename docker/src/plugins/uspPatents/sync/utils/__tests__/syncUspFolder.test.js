import 'dotenv/config';
import { test, expect } from 'vitest';

import syncUspFolder from '../http/syncUspFolder';

test('syncUSPFolder 2001', async () => {
  const destination = './data/';

  const source = `${process.env.USP_SOURCE}`;
  const year = '2001';
  let filesDownloaded = await syncUspFolder(source, destination, year);
  expect(filesDownloaded.length).toBeGreaterThan(0);
  expect(filesDownloaded).toMatchSnapshot();
}, 100000);
