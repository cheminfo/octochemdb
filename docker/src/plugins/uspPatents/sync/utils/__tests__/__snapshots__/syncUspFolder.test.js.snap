// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncUSPFolder 2001 1`] = `
[
  {
    "epoch": 978307200000,
    "href": "pa010315.zip",
    "name": "pa010315.zip",
    "path": "data/pa010315.zip",
    "size": 738265,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010315.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010322.zip",
    "name": "pa010322.zip",
    "path": "data/pa010322.zip",
    "size": 182253,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010322.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010329.zip",
    "name": "pa010329.zip",
    "path": "data/pa010329.zip",
    "size": 370579,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010329.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010405.zip",
    "name": "pa010405.zip",
    "path": "data/pa010405.zip",
    "size": 1809866,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010405.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010412.zip",
    "name": "pa010412.zip",
    "path": "data/pa010412.zip",
    "size": 1031599,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010412.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010419.zip",
    "name": "pa010419.zip",
    "path": "data/pa010419.zip",
    "size": 1216570,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010419.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010426.zip",
    "name": "pa010426.zip",
    "path": "data/pa010426.zip",
    "size": 2637670,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010426.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010503.zip",
    "name": "pa010503.zip",
    "path": "data/pa010503.zip",
    "size": 4211470,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010503.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010510.zip",
    "name": "pa010510.zip",
    "path": "data/pa010510.zip",
    "size": 5012307,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010510.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010517.zip",
    "name": "pa010517.zip",
    "path": "data/pa010517.zip",
    "size": 2305649,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010517.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010524.zip",
    "name": "pa010524.zip",
    "path": "data/pa010524.zip",
    "size": 7194151,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010524.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010531.zip",
    "name": "pa010531.zip",
    "path": "data/pa010531.zip",
    "size": 8087160,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010531.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010607.zip",
    "name": "pa010607.zip",
    "path": "data/pa010607.zip",
    "size": 8961148,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010607.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010614.zip",
    "name": "pa010614.zip",
    "path": "data/pa010614.zip",
    "size": 8283298,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010614.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010621.zip",
    "name": "pa010621.zip",
    "path": "data/pa010621.zip",
    "size": 11172693,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010621.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010628.zip",
    "name": "pa010628.zip",
    "path": "data/pa010628.zip",
    "size": 13420237,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010628.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010705.zip",
    "name": "pa010705.zip",
    "path": "data/pa010705.zip",
    "size": 16550026,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010705.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010712.zip",
    "name": "pa010712.zip",
    "path": "data/pa010712.zip",
    "size": 11665858,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010712.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010719.zip",
    "name": "pa010719.zip",
    "path": "data/pa010719.zip",
    "size": 13110978,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010719.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010726.zip",
    "name": "pa010726.zip",
    "path": "data/pa010726.zip",
    "size": 14194330,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010726.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010802.zip",
    "name": "pa010802.zip",
    "path": "data/pa010802.zip",
    "size": 18159539,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010802.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010809.zip",
    "name": "pa010809.zip",
    "path": "data/pa010809.zip",
    "size": 22763894,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010809.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010816.zip",
    "name": "pa010816.zip",
    "path": "data/pa010816.zip",
    "size": 24866729,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010816.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010823.zip",
    "name": "pa010823.zip",
    "path": "data/pa010823.zip",
    "size": 26776679,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010823.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010830.zip",
    "name": "pa010830.zip",
    "path": "data/pa010830.zip",
    "size": 24893205,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010830.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010906.zip",
    "name": "pa010906.zip",
    "path": "data/pa010906.zip",
    "size": 21389203,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010906.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010913.zip",
    "name": "pa010913.zip",
    "path": "data/pa010913.zip",
    "size": 24144540,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010913.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010920.zip",
    "name": "pa010920.zip",
    "path": "data/pa010920.zip",
    "size": 19765357,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010920.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa010927.zip",
    "name": "pa010927.zip",
    "path": "data/pa010927.zip",
    "size": 26315899,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa010927.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011004.zip",
    "name": "pa011004.zip",
    "path": "data/pa011004.zip",
    "size": 26980557,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011004.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011011.zip",
    "name": "pa011011.zip",
    "path": "data/pa011011.zip",
    "size": 29708477,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011011.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011018.zip",
    "name": "pa011018.zip",
    "path": "data/pa011018.zip",
    "size": 38546326,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011018.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011025.zip",
    "name": "pa011025.zip",
    "path": "data/pa011025.zip",
    "size": 36274060,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011025.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011101.zip",
    "name": "pa011101.zip",
    "path": "data/pa011101.zip",
    "size": 34800916,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011101.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011108.zip",
    "name": "pa011108.zip",
    "path": "data/pa011108.zip",
    "size": 31648320,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011108.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011115.zip",
    "name": "pa011115.zip",
    "path": "data/pa011115.zip",
    "size": 34449735,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011115.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011122.zip",
    "name": "pa011122.zip",
    "path": "data/pa011122.zip",
    "size": 38702761,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011122.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011129.zip",
    "name": "pa011129.zip",
    "path": "data/pa011129.zip",
    "size": 35054750,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011129.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011206.zip",
    "name": "pa011206.zip",
    "path": "data/pa011206.zip",
    "size": 29637039,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011206.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011213.zip",
    "name": "pa011213.zip",
    "path": "data/pa011213.zip",
    "size": 32647066,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011213.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011220.zip",
    "name": "pa011220.zip",
    "path": "data/pa011220.zip",
    "size": 26495427,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011220.zip",
  },
  {
    "epoch": 978307200000,
    "href": "pa011227.zip",
    "name": "pa011227.zip",
    "path": "data/pa011227.zip",
    "size": 33373144,
    "url": "https://bulkdata.uspto.gov/data/patent/application/redbook/fulltext/2001/pa011227.zip",
  },
]
`;
